### ISSUE TITLE
CodeBlock Error Underlines and Highlights

### ISSUE THREAD

**Issue by hhenrichsen** [motion-canvas/motion-canvas#258]
**Description**
Posting on behalf of two users with similar requests on the Discord; it would be useful within the code block to be able to mark something as an error and get red squiggles, as well as the option to draw a background behind certain ranges of text as an alternative to the selection logic.

**Proposed solution**
It's probably worth making the selection rendering generic enough that we can draw other types of decorations over the code, too, and then hooking into that flow for error underlines and highlights.

**Considered alternatives**
* Render a rect behind the text / an underline near the desired text -- this is workable, but doesn't update with the text and requires figuring out the coordinates of text rendering.

**Comment by aarthificial** [comment]
Before anyone starts implementing this, I'd like to see a proposal for the API that would be used for controlling this functionality.
As well as how it would play with the current `selection` property, especially tweening between different selections/highlights.

**Comment by hhenrichsen** [comment]
Fairly similar to how it was outlined in the other issue, maybe with another step between to simplify available completion options.
```ts
const codeBlock = makeRef<CodeBlock>();
// block creation...

// error - squiggled underline; lerps from left to right
codeBlock.decoration.errors(range(0, 0, 0, 10), 1) // Draws a squiggly line from left to right lerp in, same for lerp out
codeBlock.decoration.errorColor("#BF616A")

// emphasis - rename current selection to emphasis; lerps to fade in
// I get this is a breaking change so I get not renaming but instead adding this interface 
// and maybe deprecating the other name.
codeBlock.decoration.emphasis(range(0, 0, 0, 10), 1)
 // Currently the `selectionOpacity` signal seems to set the selected opacity, but it's 
// actually setting the un-selected opacity.
codeBlock.decoration.unemphasizedOpacity(0.33)

// highlight - draws a rectangle behind the text; lerps from left to right
codeBlock.decoration.highlight(range(0, 0, 0, 10), 1)
codeBlock.decoration.highlightColor("#EBCB8B77")

// underline - similar to error, but no squiggles; lerps from left to right
codeBlock.decoration.underline(range(0, 0, 0, 10), 1)
codeBlock.decoration.underlineColor("#ECEFF4")
```

**Comment by aarthificial** [comment]
I think we could try going for a more flexible solution since there's clearly a pattern there.
What about extracting the selection logic to a separate class that could completely control how a character is drawn:
```ts
const selection = new Selection((
  context, // rendering context
  char, // the character to draw
  position, // the position at which to draw the character
  rect, // a bounding box
  contained // whether the character is inside the selection
  // whatever else will be necessary
) => {
  // draw the character however you want
});

const code = <CodeBlock selections={[selection]} />
```
`CodeBlock` could then accept multiple selections like this. And we'd provide some default implementations in a form of `Highlight`, `Emphasis`, and `ErrorHighlight` classes.

**Comment by hhenrichsen** [comment]
Yeah, I was planning on building this flexible anyways. I want to try to provide a fairly simple user API to this like above, so that we don't need to do something like this:
```ts
const code = <CodeBlock selections={[new Highlight(range(0, 0, 0, 10), new Emphasis(range(0, 0, 0, 10)]}>
```
but can instead hide that inside of the CodeBlock class. I'm okay having those classes existing as a way to add decorations to code or text, but I'd like those to be easier on the user to add.

**Comment by aarthificial** [comment]
I didn't notice the `.decoration` part to be honest.
I think a simple api could look like this:
```ts
yield* code().highlight.selection(range(0, 0, 0, 10), 2);
```

Where `highlight` is a getter that checks if a `Highlight` class has already been added to the `selections` array and either retrieves it or creates a new one.

This would be consistent with the filter api, where filter are also defined using arrays:
```tsx
<Node filter={[blur(20)]}/>
```
But they provide shortcut properties for accessing any type of filter:
```ts
// this will add `saturate` to the `filters` array if one doesn't exist
// otherwise it will modify the current `saturate` filter:
node().filters.saturate(0);
```

**Comment by Ross-Esmond** [comment]
I think both of you are on the same page about having both a low level API like aarthificial showed, and _some_ high level api with more ready made classes. I think it would be easiest to open a pull for the low level first, and then introduce stuff like errors and underlines in a second pull request, just to keep things moving. Otherwise you might have to rewrite all your high level code because the core api had to change.

As for the low level API, I think the selection should give the whole string at once, rather than each character individually. There are a lot of different styles that don't work if you're only getting the individual characters. A good example being the old Motion Canvas code selection. The old code component used to draw rounded blocks around the selected code, but you can't do that rounding if you only have the individual characters. Ultimately, getting the whole string is more robust than getting just the characters, since you can always render each character independently in a loop, but you can't do it the other way around.
